akka-stream/src/main/scala/akka/stream/impl/package.scala

```diff
@@ -0,0 +1,461 @@
+ *  in the global slot-space.
+ *
+ *  == Materialization ==
+ *
+ *  Materialzation is orchestrated by the [[akka.stream.impl.PhasedFusingActorMaterializer]]. It basically decodes the
+ *  traversal and handles islands. This top-level materializer does not really handle the wiring _inside_ an island,
+ *  it only handles wiring of Publishers and Subscribers that connect islands. Instead it delegates in-island wiring
+ *  to [[akka.stream.impl.PhaseIsland]]s. For example a default fused island will be actually wired by
+ *  [[akka.stream.impl.GraphStageIsland]].
+ *
+ *  First, look at a traversal that has two islands:
+ *
+ *    |----Island1-----|----Island2(enclosed)----|-----Island1-----|
+ *
+ *  In this traversal, we have two islands, and three, so called _segments_. Segments are simply contiguous range of
+ *  slots between [[akka.stream.impl.EnterIsland]] or [[akka.stream.impl.ExitIsland]] tags (in any combination). When
+ *  the materializer encounters either an enter or exit command, it saves various information about the segment it
+ *  just completed (what is its offset, how long it is measured in input slots, etc.). This information is later
+ *  used to figure out if a wiring crosses island boundaries or is it local to the island.
+ *
+ *  It is important to note that the data structure for this is only allocated when there are islands. This is again
+ *  the 90% rule in action. In addition, these data structures are [[java.util.ArrayList]] instances, where lookups
+ *  according to some value are implemented as simple linear scans. Since in 90% of the cases these structures are
+ *  very short, this is the most efficient approach. Cases where this can be a performance problem are very-very special
+ *  and likely not happen in practice (no graph should contain more than a dozen of islands for example).
+ *
+ *  When it comes to deciding whether a wiring is cross-island or local, there are two cases possible:
+ *
+ *   * we encountered an output port that is wired backwards (relative address is negative). In this case we already
+ *     have all the data necessary to resolve the question.
+ *   * we encountered an output port that is wired forward (relative address is positive). In this case we have not
+ *     yet visited that part of the traversal where the assignment points.
+ *
+ *  If we want to keep the one-pass design of the materializer, we need to delay forward wirings until we have all
+ *  the information needed, i.e. we visit the target in port. The [[akka.stream.impl.PhasedFusingActorMaterializer]]
+ *  has a data structure for tracking forward wires which it consults whenever it visits an input port. Again, this
+ *  is only allocated if needed, and it is again an array with linear scan lookup. Once the target input port have
+ *  been found, the rules of the wiring are the same as for backwards wiring.
```

Comment:
well, at least similar. The point is we have all the information now as we have visited all the parts of the traversal that is needed to resolve the wiring.